Rubberduck.Core/UnitTesting/TestEngine.cs

```diff
@@ -88,67 +95,115 @@ public void Run(IEnumerable<TestMethod> tests)
         }
 
         private void RunInternal(IEnumerable<TestMethod> tests)
+        {
+            if (!AllowedRunStates.Contains(_state.Status))
+            {
+                return;
+            }
+
+            _state.OnSuspendParser(this, () => RunWhileSuspended(tests));
```

Comment:
From discussion w/ @MDoerner  we need to enhance this a bit more. We need to pass in an array of allowed parser states. That way when write lock is held, we should check the current status and verify we are still in the one of allowable parser state to ensure that we are not in a wrong state (e.g. a parse failed and we are in an error state before we obtained the write lock). When we determine that we cannot suspend due to wrong state, we should communicate back to the requestor that the busy action was declined.